File: lib/registrant.rb
class Registrant
    attr_reader :name,
                :age,
                :permit
                
    attr_accessor :license_data

    def initialize(name, age, permit = false)
        if name.class != String
            @name = "Default Text"
        elsif age.class != Integer
            @age = 0
        else
            @name = name
            @age = age
            @permit = permit
            @license_data = {:written=>false, :license=>false, :renewed=>false}
        end
    end

    def earn_permit
        if @permit == false && @age > 15
            @permit = true
        end
    end

    def permit?
        @permit
    end
end

File: the_dmv_extension_js/dmv/exercises/registrant.js
module.exports = {
  createRegistrant,
  earnPermit,
  takeLicenseExam
}

function createRegistrant(name, age, permit = false) {
  return {name, age, permit, licenseData: {written: false, license: false, renewed: false}}
}

function earnPermit(registrant) {
  if (!registrant.permit && registrant.age >= 16) {
    registrant.permit = true
    return 'You have successfully earned a permit!'
  }
  else {
    return 'Sorry, you are not able to earn a permit right now.'
  }
}

function takeLicenseExam(registrant) {
  if (registrant.permit) {
    registrant.licenseData.written = true
    return 'Great job! You can take your test.'
  }
  else {
    return 'Sorry, you need to earn a permit first.'
  }
}
